lib/propolis/src/hw/qemu/pvpanic.rs

```diff
@@ -0,0 +1,80 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::sync::{
+    atomic::{AtomicUsize, Ordering::Relaxed},
+    Arc,
+};
+
+use crate::common::*;
+use crate::pio::{PioBus, PioFn};
+
+pub struct QemuPioPvpanic {
+    counts: Arc<PanicCounts>,
+}
+
+#[derive(Debug, Default)]
+pub struct PanicCounts {
+    host_handled: AtomicUsize,
+    guest_handled: AtomicUsize,
+}
+
+/// Indicates that a guest panic has happened and should be processed by the
+/// host
+const HOST_HANDLED: u8 = 0b01;
+/// Indicates a guest panic has happened and will be handled by the guest; the
+/// host should record it or report it, but should not affect the execution of
+/// the guest.
+const GUEST_HANDLED: u8 = 0b10;
+
+impl QemuPioPvpanic {
+    const IOPORT: u16 = 0x505;
+
+    #[must_use]
+    pub fn create(pio: &PioBus, counts: Arc<PanicCounts>) -> Arc<Self> {
+        let this = Arc::new(Self { counts });
+
+        let piodev = this.clone();
+        let piofn = Arc::new(move |_port: u16, rwo: RWOp| piodev.pio_rw(rwo))
+            as Arc<PioFn>;
+        pio.register(Self::IOPORT, 1, piofn).unwrap();
+        this
+    }
+
+    fn pio_rw(&self, rwo: RWOp) {
+        match rwo {
+            RWOp::Read(ro) => {
+                ro.write_u8(HOST_HANDLED | GUEST_HANDLED);
+            }
+            RWOp::Write(wo) => {
+                let c = wo.read_u8();
+                if c & HOST_HANDLED != 0 {
+                    self.counts.host_handled.fetch_add(1, Relaxed);
+                }
+
+                if c & GUEST_HANDLED != 0 {
+                    self.counts.guest_handled.fetch_add(1, Relaxed);
+                }
+            }
+        }
+    }
+}
+
+impl PanicCounts {
+    #[must_use]
```

Comment:
Why the `must_use` when the consequences for ignoring it are basically nil?

Reply:
The style I'm used to from other projects is to put `#[must_use]` on most functions that return a value without performing any side effects (e.g. accessors, constructors). There are no _consequences_ for ignoring the return value from these functions, but these functions are called only to produce the return value --- if  the return value is not used, the function may as well not have been called, so calling them without using the value is generally a mistake. 

I'm happy to remove the `#[must_use]` attributes if this style is controversial, though.

Reply:
I'm inclined to elide it, for now, in this kind of case.  Maybe we'll decide that style is a fit for propolis in the future.  That said, there are probably a bunch of existing cases where ignoring a return value will cause correctness issues, so please do not hesitate to point those out.  (I've been trying to fix some of them as time passes)